src/dmd/clone.d

```diff
@@ -883,6 +884,34 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
             e = Expression.combine(ex, e); // combine in reverse order
         }
 
+        /* extern(C++) destructors call into super to destruct the full hierarchy
+        */
+        ClassDeclaration cldec = ad.isClassDeclaration();
+        if (cldec && cldec.classKind == ClassKind.cpp && cldec.baseClass && cldec.baseClass.dtor)
+        {
+            // WAIT BUT: do I need to run `cldec.baseClass.dtor` semantic? would it have been run before?
+            cldec.baseClass.dtor.functionSemantic();
+
+            stc = mergeFuncAttrs(stc, cldec.baseClass.dtor);
+            if (!(stc & STC.disable))
+            {
+                // super.__xdtor()
+
+                Expression ex = new SuperExp(loc);
+
+                // This is a hack so we can call destructors on const/immutable objects.
+                // Do it as a type 'paint'.
+                ex = new CastExp(loc, ex, cldec.baseClass.type.mutableOf());
+                if (stc & STC.safe)
+                    stc = (stc & ~STC.safe) | STC.trusted;
+
+                ex = new DotVarExp(loc, ex, cldec.baseClass.dtor, false);
+                ex = new CallExp(loc, ex);
```

Comment:
Would be nice if you could forward the dtor argument here, so it doesn't get lost when `delete` is called from C++ and the base class is also in C++ and handles it as expected.

Reply:
If C++ ever calls a destructor with non-0, it means an operator delete override exists beneath it in the hierarchy, and therefore is not supported by D.
There is no case where it's valid that any argument other than 0 is used, and D is also involved. I dont think 'we can support this, so I'm not sure fixing this is worth the complexity no?

Reply:
Also, I think it's only the top-level destructor that calls delete, you don't pass the request to delete down the hierarchy, otherwise it would call delete at each level of destruction... only the top-level destruction should finally delete.

Reply:
Right. If we add the enhancement to support calling C++ delete, then we should be semantically compatible I think.